dali/operators/reader/numpy_reader_gpu_op_impl.cu

```diff
@@ -0,0 +1,164 @@
+  bool need_slice = !rois_.empty();
+  auto out_view = view<T>(output);
+  auto curr_batch = GetCurrBatchView<T, Dims>();
+  auto dtype = TypeTable::GetTypeInfoFromStatic<T>();
+
+  // Permuted dims, to use for the transposition
+  std::array<int, Dims> perm;
+  for (int d = 0; d < Dims; d++)
+    perm[d] = Dims - 1 - d;
+
+  int nsamples_copy = 0;
+  for (int i = 0; i < nsamples; i++)
+    if (!need_slice_[i] && !need_transpose_[i])
+      nsamples_copy++;
+  if (nsamples_copy) {
+    if (nsamples_copy == nsamples) {
+      std::swap(output, prefetched_batch_tensors_[curr_batch_consumer_]);
+      return;
+    }
+    for (int i = 0; i < nsamples; i++) {
+      if (need_slice_[i] || need_transpose_[i])
+        continue;
+      auto sz = out_sh.tensor_size(i) * dtype.size();
+      sg_.AddCopy(out_view.data[i], curr_batch.data[i], sz);
+    }
+    sg_.Run(ws.stream());
+  }
+
+  int nsamples_slice = std::count(need_slice_.begin(), need_slice_.end(), true);
+  int nsamples_transpose = std::count(need_transpose_.begin(), need_transpose_.end(), true);
+  int nsamples_slice_transpose = 0;
+  for (int i = 0; i < nsamples; i++)
+    if (need_slice_[i] && need_transpose_[i])
+      nsamples_slice_transpose++;
+
+  TensorListView<StorageGPU, T, Dims> tmp_view;
+  if (nsamples_slice_transpose) {
+    tmp_buf_sh_.resize(nsamples_slice_transpose, Dims);
+    for (int i = 0, j = 0; i < nsamples; i++) {
+      if (need_slice_[i] && need_transpose_[i]) {
+        tmp_buf_sh_.set_tensor_shape(j, rois_[i].shape);
+        j++;
+      }
+    }
+    tmp_buf_.Resize(tmp_buf_sh_, dtype);
+    tmp_view = view<T, Dims>(tmp_buf_);
+  }
+
+  if (nsamples_slice) {
+    // TLV used to invoke kernels with a subset of the samples
+    TensorListView<StorageGPU, const T, Dims> from;
+    from.resize(nsamples_slice);
+    TensorListView<StorageGPU, T, Dims> to;
+    to.resize(nsamples_slice);
+
+    std::vector<kernels::SliceArgs<T, Dims>> slice_args;
+    slice_args.resize(nsamples_slice);
+    for (int i = 0, j = 0, k = 0; i < nsamples; i++) {
+      if (!need_slice_[i])
+        continue;
+      auto &args = slice_args[j];
+      args.anchor = rois_[i].anchor;
+      args.shape = rois_[i].shape;
+      args.fill_values.clear();
+      args.fill_values.push_back(ConvertSat<T>(fill_value_));
+
+      from.data[j] = curr_batch.data[i];
+      from.shape.set_tensor_shape(j, curr_batch.shape[i]);
+
+      if (need_transpose_[i]) {
+        // slice to an intermediate buffer
+        to.data[j] = tmp_view.data[k];
+        to.shape.set_tensor_shape(j, tmp_view.tensor_shape(k));
+        k++;
+      } else {
+        // directly to the output
+        to.data[j] = out_view.data[i];
+        to.shape.set_tensor_shape(j, out_view.shape[i]);
+      }
+      j++;
+    }
+    kernels::KernelContext ctx;
+    ctx.gpu.stream = ws.stream();
+    using Kernel = kernels::SliceGPU<T, T, Dims>;
+    kmgr_slice_.Resize<Kernel>(1, 1);
+    kmgr_slice_.Setup<Kernel>(0, ctx, from, slice_args);
```

Comment:
Does it make sense to do this Setup in SetupImpl?

Reply:
Not really. We already know the size, so we can run setup just before run. It saves boilerplate (value switch, type switch, etc)

Reply:
Sure